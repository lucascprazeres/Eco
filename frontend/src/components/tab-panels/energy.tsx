import { CarbonFootprintInput, TabsEnum } from '@eco/models/carbon-footprint'
import { TabPanel } from '@mui/lab'
import { Box, Button, TextField, Typography } from '@mui/material'
import { useFormContext } from 'react-hook-form'

interface EnergyPanelProps {
  onClickNext: () => void
}

export function EnergyPanel({ onClickNext }: EnergyPanelProps) {
  const { register, handleSubmit, formState } =
    useFormContext<CarbonFootprintInput>()

  function handleGoToNextTab() {
    onClickNext()
  }

  const error = formState.errors?.electricityUsageKWhPerMonth?.message

  return (
    <TabPanel
      value={TabsEnum.EnergyUsage}
      style={{ height: '100%', width: '100%' }}
    >
      <form
        onSubmit={handleSubmit(handleGoToNextTab)}
        style={{ display: 'flex', flexDirection: 'column', height: '100%' }}
      >
        <Box
          width="100%"
          height="100%"
          display="flex"
          flexDirection="column"
          gap={4}
        >
          <Typography
            variant="h6"
            fontWeight="bold"
            fontFamily="ubuntu"
            color="secondary.dark"
          >
            Enter your monthly electricity usage
          </Typography>

          <TextField
            {...register('electricityUsageKWhPerMonth', {
              valueAsNumber: true,
              required: {
                value: true,
                message: 'Please insert your electric usage',
              },
            })}
            label="Electricity usage (Kwh/month)"
            type="number"
            fullWidth
            helperText={error}
            error={!!error}
          />

          <Box
            display="flex"
            alignSelf="flex-end"
            alignItems="center"
            justifyContent="flex-end"
            width="100%"
            bottom={0}
          >
            <Button
              type="submit"
              color="primary"
              variant="contained"
              disabled={!!error}
              sx={{ fontWeight: 'bold' }}
            >
              Next
            </Button>
          </Box>
        </Box>
      </form>
    </TabPanel>
  )
}
