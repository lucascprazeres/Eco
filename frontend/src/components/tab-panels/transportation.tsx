import { CarbonFootprintInput } from '@eco/models/carbon-footprint'
import {
  Box,
  Button,
  FormControlLabel,
  Radio,
  RadioGroup,
  TextField,
  Typography,
} from '@mui/material'
import { useController, useFormContext } from 'react-hook-form'

interface TransportationPanelProps {
  onClickNext: () => void
  onGoBack: () => void
}

export function TransportationPanel({
  onClickNext,
  onGoBack,
}: TransportationPanelProps) {
  const { register, handleSubmit, formState, control } =
    useFormContext<CarbonFootprintInput>()

  const { field, fieldState } = useController({
    name: 'transportationFuelType',
    control,
    rules: {
      required: {
        value: true,
        message: 'Please select at least one option',
      },
    },
  })

  function handleGoToNextTab() {
    onClickNext()
  }

  const error = formState.errors?.transportationFuelGallonsPerMonth?.message

  return (
    <form
      onSubmit={handleSubmit(handleGoToNextTab)}
      style={{ display: 'flex', flexDirection: 'column', height: '100%' }}
    >
      <Box
        width="100%"
        height="100%"
        display="flex"
        flexDirection="column"
        gap={4}
      >
        <Box>
          <Typography variant="h6" fontWeight="bold" fontFamily="ubuntu">
            What type of fuel does your vehicle use?
          </Typography>

          <RadioGroup
            name="transporationFuelType"
            value={field.value}
            onChange={field.onChange}
          >
            <Box display="flex" alignItems="center" gap={1}>
              <FormControlLabel
                value="gasoline"
                label="Gasoline"
                control={<Radio />}
              />
              <FormControlLabel
                label="Diesel"
                value="diesel"
                control={<Radio />}
              />
              <FormControlLabel
                label="Ethanol"
                value="ethanol"
                control={<Radio />}
              />
            </Box>
          </RadioGroup>

          <Typography variant="caption" color="error">
            {fieldState.error?.message}
          </Typography>
        </Box>

        <Typography variant="h6" fontWeight="bold" fontFamily="ubuntu">
          How many gallons of fuel do you use in a month?
        </Typography>

        <TextField
          {...register('transportationFuelGallonsPerMonth', {
            valueAsNumber: true,
            required: {
              value: true,
              message: 'Please insert your fuel usage (or 0 if none)',
            },
          })}
          label="Fuel (gallons)"
          type="number"
          style={{ width: '100%' }}
          helperText={error}
          error={!!error}
        />

        <Box
          display="flex"
          alignSelf="flex-end"
          alignItems="center"
          justifyContent="space-between"
          width="100%"
          bottom={0}
        >
          <Button variant="outlined" onClick={onGoBack}>
            Go back
          </Button>
          <Button
            type="submit"
            color="primary"
            variant="contained"
            disabled={!!error}
          >
            Next
          </Button>
        </Box>
      </Box>
    </form>
  )
}
