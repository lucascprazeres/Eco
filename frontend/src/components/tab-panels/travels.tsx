import { RESULTS_PAGE } from '@eco/constants/routes'
import { CarbonFootprintInput } from '@eco/models/carbon-footprint'
import { Box, Button, TextField, Typography } from '@mui/material'
import { useRouter } from 'next/router'
import { useCallback } from 'react'
import { useFormContext } from 'react-hook-form'

interface TravelPanelProps {
  onGoBack: () => void
  onSubmit: (data: CarbonFootprintInput) => Promise<void>
}

export function TravelPanel({ onGoBack, onSubmit }: TravelPanelProps) {
  const router = useRouter()
  const { register, handleSubmit, formState } =
    useFormContext<CarbonFootprintInput>()

  const handleLastSubmit = useCallback(
    async (data: CarbonFootprintInput) => {
      try {
        await onSubmit(data)
        router.push(RESULTS_PAGE)
      } catch (err) {
        console.log(err)
      }
    },
    [router, onSubmit],
  )

  const travelsPerYearError = formState.errors?.airTravelsPerYear?.message

  return (
    <form
      onSubmit={handleSubmit(handleLastSubmit)}
      style={{ display: 'flex', flexDirection: 'column', height: '100%' }}
    >
      <Box
        width="100%"
        height="100%"
        display="flex"
        flexDirection="column"
        gap={4}
      >
        <Typography
          variant="h6"
          fontWeight="bold"
          fontFamily="ubuntu"
          color="secondary.dark"
        >
          How many flights do you take per year?
        </Typography>

        <TextField
          {...register('airTravelsPerYear', {
            valueAsNumber: true,
            required: {
              value: true,
              message:
                'Please insert how many flights do you take (or 0 if none)',
            },
          })}
          label="Flights (per year)"
          type="number"
          fullWidth
          helperText={travelsPerYearError}
          error={!!travelsPerYearError}
        />

        <Box
          display="flex"
          alignSelf="flex-end"
          alignItems="center"
          justifyContent="space-between"
          width="100%"
          bottom={0}
        >
          <Button
            variant="outlined"
            onClick={onGoBack}
            sx={{ fontWeight: 'bold' }}
          >
            Go back
          </Button>
          <Button
            type="submit"
            color="primary"
            variant="contained"
            disabled={!!travelsPerYearError}
            sx={{ fontWeight: 'bold' }}
          >
            Next
          </Button>
        </Box>
      </Box>
    </form>
  )
}
