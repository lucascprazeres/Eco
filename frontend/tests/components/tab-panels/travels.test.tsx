import { describe, it, expect, vi, afterEach, Mock } from 'vitest'
import {
  render,
  screen,
  fireEvent,
  cleanup,
  waitFor,
} from '@testing-library/react'
import { TravelPanel } from '@eco/components/tab-panels/travels'
import { useFormContext } from 'react-hook-form'
import { useFormContextMockReturn } from '../../mocks/react-hook-form'
import mockRouter from 'next-router-mock'

vi.mock('next/router', () => vi.importActual('next-router-mock'))

vi.mock('react-hook-form', () => ({
  useFormContext: vi.fn(() => useFormContextMockReturn),
}))

describe('TravelPanel', () => {
  afterEach(() => {
    cleanup()
    vi.clearAllMocks()
  })

  it('should be able to render correctly', () => {
    // arrange
    const mockOnGoBack = vi.fn()
    const mockOnSubmit = vi.fn()

    // act
    render(<TravelPanel onGoBack={mockOnGoBack} onSubmit={mockOnSubmit} />)

    const title = screen.getByText(/How many flights do you take per year\?/i)
    const input = screen.getByLabelText(/Flights \(per year\)/i)
    const backButton = screen.getByRole('button', { name: /Go back/i })
    const nextButton = screen.getByRole('button', { name: /Next/i })

    // assert
    expect(title).toBeInTheDocument()
    expect(input).toBeInTheDocument()
    expect(backButton).toBeInTheDocument()
    expect(nextButton).toBeInTheDocument()
    expect(nextButton).not.toBeDisabled()
  })

  it('should be able to display validation error for invalid input', () => {
    // arrange
    // eslint-disable-next-line prettier/prettier
    (useFormContext as Mock).mockReturnValueOnce({
      register: vi.fn(),
      handleSubmit: vi.fn(),
      formState: {
        errors: {
          airTravelsPerYear: { message: 'Validation error' },
        },
      },
    })

    // act
    render(<TravelPanel onGoBack={vi.fn()} onSubmit={vi.fn()} />)

    const helperText = screen.getByText(/Validation error/i)
    const nextButton = screen.getByRole('button', { name: /Next/i })

    // assert
    expect(helperText).toBeInTheDocument()
    expect(nextButton).toBeDisabled()
  })

  it('should call onSubmit and navigate to results page when the form is submitted without errors', async () => {
    // arrange
    const mockOnSubmit = vi.fn()

    // act
    render(<TravelPanel onGoBack={vi.fn()} onSubmit={mockOnSubmit} />)

    const nextButton = screen.getByRole('button', { name: /Next/i })
    fireEvent.submit(nextButton)

    // assert
    expect(mockOnSubmit).toHaveBeenCalled()

    await waitFor(() =>
      expect(mockRouter).toMatchObject({
        pathname: '/result',
      }),
    )
  })

  it('should call onGoBack when the back button is clicked', () => {
    // arrange
    const mockOnGoBack = vi.fn()

    // act
    render(<TravelPanel onGoBack={mockOnGoBack} onSubmit={vi.fn()} />)

    const backButton = screen.getByRole('button', { name: /Go back/i })
    fireEvent.click(backButton)

    // assert
    expect(mockOnGoBack).toHaveBeenCalled()
  })
})
