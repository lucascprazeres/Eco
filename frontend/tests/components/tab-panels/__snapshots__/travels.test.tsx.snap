// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TravelPanel > should be able to render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        style="display: flex; flex-direction: column; height: 100%;"
      >
        <div
          class="MuiBox-root css-nf1kcv"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-b5iqwz-MuiTypography-root"
          >
            How many flights do you take per year?
          </h6>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r0:"
              id=":r0:-label"
            >
              Flights (per year)
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                type="number"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-81qg8w"
                >
                  <span>
                    Flights (per year)
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiBox-root css-14i6vh8"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-uah78b-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Go back
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-r9p8f3-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Next
            </button>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      style="display: flex; flex-direction: column; height: 100%;"
    >
      <div
        class="MuiBox-root css-nf1kcv"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-b5iqwz-MuiTypography-root"
        >
          How many flights do you take per year?
        </h6>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Flights (per year)
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="number"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-81qg8w"
              >
                <span>
                  Flights (per year)
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiBox-root css-14i6vh8"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-uah78b-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Go back
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-r9p8f3-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Next
          </button>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
